File: artifacts/gauntlet-heat2-20260824/openclaw/rider-driver.mjs
#!/usr/bin/env node
import { spawn } from "node:child_process";
import { createWriteStream, mkdirSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";

const [, , contract, seed, attempt, tapePath, logPath] = process.argv;

if (!contract || !seed || !attempt || !tapePath || !logPath) {
  console.error("usage: rider-driver.mjs <contract> <seed> <attempt> <tapePath> <logPath>");
  process.exit(64);
}

mkdirSync(dirname(resolve(tapePath)), { recursive: true });
mkdirSync(dirname(resolve(logPath)), { recursive: true });

const startedAt = Date.now();
const log = createWriteStream(logPath, { flags: "w" });
const child = spawn(
  "node",
  ["scripts/gr-sim.mjs", "--contract", contract, "--seed", seed, "--tape", resolve(tapePath)],
  { cwd: "/tmp/heat2-b42c0fbc", stdio: ["pipe", "pipe", "pipe"] },
);

let stdoutBuf = "";
let stderrBuf = "";
let lineNo = 0;
let lastOutcome = null;
let terminalView = null;
let sent = 0;

child.stderr.setEncoding("utf8");
child.stderr.on("data", (chunk) => {
  stderrBuf += chunk;
  log.write(JSON.stringify({ type: "stderr", chunk }) + "\n");
});

child.stdout.setEncoding("utf8");
child.stdout.on("data", (chunk) => {
  stdoutBuf += chunk;
  for (;;) {
    const ix = stdoutBuf.indexOf("\n");
    if (ix < 0) break;
    const raw = stdoutBuf.slice(0, ix).trim();
    stdoutBuf = stdoutBuf.slice(ix + 1);
    if (!raw) continue;
    handleLine(raw);
  }
});

child.on("close", (code, signal) => {
  const wallSeconds = (Date.now() - startedAt) / 1000;
  const summary = {
    contract,
    seed,
    attempt: Number(attempt),
    code,
    signal,
    sent,
    wallSeconds,
    outcome: lastOutcome,
    terminalView,
    stderrTail: stderrBuf.slice(-4000),
  };
  log.write(JSON.stringify({ type: "summary", summary }) + "\n");
  log.end();
  writeFileSync(logPath.replace(/\.ndjson$/, ".summary.json"), JSON.stringify(summary, null, 2));
  console.log(JSON.stringify(summary));
});

function handleLine(raw) {
  lineNo += 1;
  let parsed;
  try {
    parsed = JSON.parse(raw);
  } catch (error) {
    log.write(JSON.stringify({ type: "stdout-raw", lineNo, raw }) + "\n");
    return;
  }
  log.write(JSON.stringify({ type: "stdout-json", lineNo, value: parsed }) + "\n");
  if (parsed?.schema === "goldrush.view.v1") {
    if (isTerminalView(parsed)) terminalView = summarizeView(parsed);
    const orders = makeOrders(parsed);
    child.stdin.write(JSON.stringify(orders) + "\n\n");
    sent += 1;
    log.write(JSON.stringify({ type: "orders", sent, orders }) + "\n");
  } else {
    lastOutcome = parsed;
  }
}

function isTerminalView(view) {
  return Boolean(view?.now?.pendingSecure || view?.now?.terminal || view?.now?.outcome);
}

function summarizeView(view) {
  return {
    wave: view.now?.wave,
    hp: view.now?.hero?.hp,
    gold: view.now?.gold,
    pendingSecure: view.now?.pendingSecure ?? null,
    threats: view.now?.threats ?? null,
    score: view.now?.score ?? null,
  };
}

function makeOrders(view) {
  const now = view.now ?? {};
  if (now.pendingSecure) return [{ verb: "SECURE_CHOICE", choice: "bank" }];
  if (Array.isArray(now.pendingOffer) && now.pendingOffer.length) {
    return makeContractOrders(view);
  }

  return makeContractOrders(view);
}

function makeContractOrders(view) {
  if (contract === "e2-hill-mine") return hillMineOrders(view);
  if (contract === "e1-night-shift") return nightShiftOrders(view);
  return claimOrders(view);
}

function pickUpgrade(offers, now = {}) {
  const ranked = [...offers].sort((a, b) => upgradeScore(b, now) - upgradeScore(a, now));
  return ranked[0] ?? offers[0];
}

function upgradeScore(offer, now = {}) {
  const hay = `${offer.id ?? ""} ${offer.name ?? ""} ${offer.effectText ?? ""}`.toLowerCase();
  let score = 0;
  const hurt = (now.hero?.hp ?? 100) < Math.max(75, (now.hero?.maxHp ?? 100) * 0.75);
  if (hurt && (hay.includes("heal") || hay.includes("health") || hay.includes("hp") || hay.includes("plating"))) score += 180;
  if (hay.includes("turret")) score += 80;
  if (hay.includes("spark") || hay.includes("rig")) score += 70;
  if (hay.includes("blast")) score += 65;
  if (hay.includes("damage")) score += 50;
  if (hay.includes("range")) score += 45;
  if (hay.includes("cooldown")) score += 42;
  if (hay.includes("repair")) score += 40;
  if (hay.includes("slow") || hay.includes("sentry")) score += 35;
  if (hay.includes("seam") || hay.includes("gold") || hay.includes("harvest")) score += 25;
  if (hay.includes("health") || hay.includes("hp")) score += 20;
  return score;
}

function claimOrders(view) {
  const claim = view.stablePrefix.map.claim;
  const station = { x: claim.x, z: claim.z - 1 };
  const desired = [
    build("palisade", claim.x, claim.z - 4),
    build("palisade", claim.x + 4, claim.z),
    build("palisade", claim.x - 4, claim.z),
    build("sentry_beacon", claim.x, claim.z - 8),
    build("turret", claim.x, claim.z - 2),
    build("palisade", claim.x, claim.z + 4),
    build("sentry_beacon", claim.x + 8, claim.z),
    build("sentry_beacon", claim.x - 8, claim.z),
    build("palisade", claim.x + 4, claim.z - 4),
    build("palisade", claim.x - 4, claim.z - 4),
    build("turret", claim.x + 5, claim.z - 1),
    build("turret", claim.x - 5, claim.z - 1),
    build("sentry_beacon", claim.x, claim.z + 8),
    build("turret", claim.x, claim.z + 5),
  ];
  return frontierOrders(view, station, desired, { repairPct: 76, harvestBursts: 18 });
}

function nightShiftOrders(view) {
  const claim = view.stablePrefix.map.claim;
  const station = { x: claim.x, z: claim.z - 1 };
  const desired = [
    build("palisade", claim.x, claim.z - 4),
    build("palisade", claim.x + 4, claim.z),
    build("palisade", claim.x - 4, claim.z),
    build("turret", claim.x, claim.z - 2),
    build("palisade", claim.x, claim.z + 4),
    build("turret", claim.x + 5, claim.z - 1),
    build("turret", claim.x - 5, claim.z - 1),
    build("sentry_beacon", claim.x, claim.z - 6),
    build("sentry_beacon", claim.x + 6, claim.z),
    build("sentry_beacon", claim.x - 6, claim.z),
    build("sentry_beacon", claim.x, claim.z + 8),
  ];
  return frontierOrders(view, station, desired, { repairPct: 100, harvestBursts: 18, weapon: "rig", explicitBlast: true, kite: true });
}

function hillMineOrders(view) {
  const claim = view.stablePrefix.map.claim;
  const station = { x: claim.x, z: claim.z - 1 };
  const hasTurret = (view.now?.works?.entries ?? []).some((entry) => entry.id === "turret" && !entry.wrecked);
  if (!hasTurret) {
    const orders = [];
    if ((view.now?.threats?.alive ?? 0) > 0 && (view.now?.blastReadyInMs ?? 1) === 0) orders.push({ verb: "BLAST_AT", pos: blastPoint(view, view.now?.threats?.edge) });
    orders.push({ verb: "SET_WEAPON", weapon: "rig" });
    const seams = harvestQueue(activeSeams(view, station)).slice(0, 12);
    for (const seam of seams) orders.push({ verb: "HARVEST", seam: seam.id });
    orders.push({ verb: "MOVE_TO", pos: station });
    orders.push({ verb: "BUILD", what: "turret", where: { x: claim.x, z: claim.z - 2 }, when: { goldGte: 50 } });
    orders.push({ verb: "BUILD", what: "palisade", where: { x: claim.x, z: claim.z - 4 }, when: { goldGte: 10 } });
    orders.push({ verb: "BUILD", what: "palisade", where: { x: claim.x + 4, z: claim.z }, when: { goldGte: 10 } });
    orders.push({ verb: "HOLD", pos: station });
    return orders.slice(0, 32);
  }
  const desired = [
    build("palisade", claim.x, claim.z - 4),
    build("palisade", claim.x + 4, claim.z),
    build("palisade", claim.x - 4, claim.z),
    build("palisade", claim.x, claim.z + 4),
    build("turret", claim.x + 5, claim.z - 1),
    build("turret", claim.x - 5, claim.z - 1),
    build("turret", claim.x, claim.z + 5),
    build("sentry_beacon", claim.x, claim.z + 4),
    build("sentry_beacon", claim.x + 5, claim.z),
    build("sentry_beacon", claim.x - 5, claim.z),
    build("boiler_house", claim.x, claim.z),
  ];
  return frontierOrders(view, station, desired, { repairPct: 78, harvestBursts: 18 });
}

function frontierOrders(view, station, desired, options) {
  const now = view.now ?? {};
  const orders = [];
  const edge = now.threats?.edge;
  if (options.explicitBlast !== false && (now.threats?.alive ?? 0) > 0 && (now.blastReadyInMs ?? 1) === 0) {
    orders.push({ verb: "BLAST_AT", pos: blastPoint(view, edge) });
  }
  orders.push({ verb: "SET_WEAPON", weapon: options.weapon ?? "rig" });
  if (options.repairPct != null && shouldRepairNow(view, options.repairPct)) {
    orders.push({ verb: "REPAIR_UNDER", pct: options.repairPct });
  }

  const missing = desired.filter((target) => isBuildable(view, target.what) && !hasWorkNear(now.works?.entries, target));
  const plannedCounts = new Map();
  const seams = activeSeams(view, station);
  const combatPos = options.kite ? kitePoint(view, station) : station;
  if (((now.threats?.alive ?? 0) > 0 && now.threats?.state === "active") || (now.needsRider && (now.threats?.alive ?? 0) > 0) || ((now.hero?.hp ?? 100) < 50 && (now.threats?.alive ?? 0) > 0) || (options.kite && (now.threats?.alive ?? 0) >= 10)) {
    orders.push({ verb: "MOVE_TO", pos: combatPos });
    for (const target of combatBuilds(view, missing, station)) {
      if (orders.length >= 31) break;
      const plannedIndex = plannedCounts.get(target.what) ?? countWorks(now.works?.entries, target.what);
      plannedCounts.set(target.what, plannedIndex + 1);
      orders.push({ verb: "MOVE_TO", pos: nearBuild(target.where, station) });
      orders.push({ verb: "BUILD", what: target.what, where: target.where, when: { goldGte: costFor(view, target.what, plannedIndex) } });
    }
    orders.push({ verb: "HOLD", pos: combatPos });
    return orders.slice(0, 32);
  }

  const harvestBursts = Math.max(options.harvestBursts, missing.length * 3);
  const planned = [];
  const harvests = harvestQueue(seams);
  if (options.kite) orders.push({ verb: "FALLBACK_IF", threat: { enemiesGte: 8 }, pos: combatPos });

  for (let i = 0; i < harvestBursts && planned.length < 24; i += 1) {
      const seam = harvests[i % Math.max(harvests.length, 1)];
      if (seam) planned.push({ verb: "HARVEST", seam: seam.id });
      if (options.repairRoute && i % 4 === 1) {
        const repairTarget = repairTargets(view, options.repairRoute, station)[Math.floor(i / 4)];
        if (repairTarget) {
          planned.push({ verb: "MOVE_TO", pos: nearBuild(repairTarget.position, station) });
          planned.push({ verb: "REPAIR_UNDER", pct: 100 });
        }
      }
      if (i % 2 === 1) {
      const target = missing[Math.floor(i / 2)];
      if (target) {
        const plannedIndex = plannedCounts.get(target.what) ?? countWorks(now.works?.entries, target.what);
        plannedCounts.set(target.what, plannedIndex + 1);
        planned.push({ verb: "MOVE_TO", pos: nearBuild(target.where, station) });
        planned.push({ verb: "BUILD", what: target.what, where: target.where, when: { goldGte: costFor(view, target.what, plannedIndex) } });
      }
    }
  }

  for (const order of planned) {
    if (orders.length >= 31) break;
    orders.push(order);
  }
  orders.push({ verb: "HOLD", pos: station });
  return orders.slice(0, 32);
}

function shouldRepairNow(view) {
  const gold = view.now?.gold ?? 0;
  if (gold < 8) return false;
  const here = view.now?.prospector ?? view.now?.hero ?? { x: 0, z: 0 };
  return (view.now?.works?.entries ?? []).some((entry) => {
    const pos = entry.position ?? entry.pos ?? entry.where ?? entry;
    return (entry.hp ?? 0) < (entry.maxHp ?? 0) && dist(here, pos) <= 5;
  });
}

function repairTargets(view, id, station) {
  return (view.now?.works?.entries ?? [])
    .filter((entry) => entry.id === id && (entry.hp ?? 0) < (entry.maxHp ?? 0) && entry.position)
    .sort((a, b) => dist(a.position, station) - dist(b.position, station))
    .slice(0, 4);
}

function combatBuilds(view, missing, station) {
  const gold = view.now?.gold ?? 0;
  const priority = { turret: 5, sentry_beacon: 4, lantern_post: 3, boiler_house: 3, palisade: 2, sluice: 1 };
  return [...missing]
    .sort((a, b) => (priority[b.what] ?? 0) - (priority[a.what] ?? 0))
    .filter((target) => dist(target.where, station) <= 6)
    .filter((target) => costFor(view, target.what) <= gold)
    .slice(0, 2);
}

function nearBuild(where, station) {
  const d = dist(where, station);
  if (d <= 2.5) return station;
  const t = 2.5 / d;
  return {
    x: round2(where.x + (station.x - where.x) * t),
    z: round2(where.z + (station.z - where.z) * t),
  };
}

function build(what, x, z) {
  return { what, where: { x: round2(x), z: round2(z) } };
}

function round2(n) {
  return Math.round(n * 100) / 100;
}

function isBuildable(view, id) {
  return (view.stablePrefix?.mechanics?.buildables ?? []).some((b) => b.id === id);
}

function countWorks(entries = [], id) {
  return entries.filter((entry) => (entry.id ?? entry.kind ?? entry.what) === id).length;
}

function costFor(view, id, index = countWorks(view.now?.works?.entries, id)) {
  const meta = (view.stablePrefix?.mechanics?.buildables ?? []).find((b) => b.id === id);
  if (!meta) return 10;
  if (meta.costs?.[index] != null) return meta.costs[index];
  const last = meta.costs?.[meta.costs.length - 1] ?? meta.cost ?? 10;
  return last;
}

function hasWorkNear(entries = [], target) {
  return entries.some((entry) => {
    const id = entry.id ?? entry.kind ?? entry.what;
    const pos = entry.position ?? entry.pos ?? entry.where ?? entry;
    return id === target.what && dist(pos, target.where) <= 1.25;
  });
}

function activeSeams(view, point) {
  const seams = (view.now?.seams ?? []).filter((seam) => seam.active && seam.remaining > 0 && typeof seam.x === "number" && typeof seam.z === "number");
  seams.sort((a, b) => dist(a, point) - dist(b, point));
  return seams;
}

function harvestQueue(seams) {
  const out = [];
  for (const seam of seams) {
    const count = Math.max(1, Math.floor((seam.remaining ?? 5) / 5));
    for (let i = 0; i < count; i += 1) out.push(seam);
  }
  return out;
}

function dist(a, b) {
  const ax = a?.x ?? 0;
  const az = a?.z ?? 0;
  const bx = b?.x ?? 0;
  const bz = b?.z ?? 0;
  return Math.hypot(ax - bx, az - bz);
}

function blastPoint(view, edge) {
  const h = view.now?.hero ?? view.now?.prospector ?? { x: 0, z: 0 };
  const gate = gateVector(view, edge, h);
  if (gate) return { x: round2((h.x ?? 0) + gate.x * 8), z: round2((h.z ?? 0) + gate.z * 8) };
  const offsets = {
    north: { x: 0, z: -8 },
    south: { x: 0, z: 8 },
    east: { x: 8, z: 0 },
    west: { x: -8, z: 0 },
  };
  const o = offsets[edge] ?? { x: 0, z: 0 };
  return { x: round2((h.x ?? 0) + o.x), z: round2((h.z ?? 0) + o.z) };
}

function kitePoint(view, station) {
  const h = view.now?.hero ?? view.now?.prospector ?? station;
  const edge = view.now?.threats?.edge;
  const gate = gateVector(view, edge, h);
  if (gate) return { x: round2((h.x ?? station.x) - gate.x * 9), z: round2((h.z ?? station.z) - gate.z * 9) };
  const offsets = {
    north: { x: 0, z: 9 },
    south: { x: 0, z: -9 },
    east: { x: -9, z: 0 },
    west: { x: 9, z: 0 },
  };
  const o = offsets[edge] ?? { x: 0, z: -7 };
  return { x: round2((h.x ?? station.x) + o.x), z: round2((h.z ?? station.z) + o.z) };
}

function gateVector(view, edge, from) {
  if (!edge) return null;
  const gates = (view.stablePrefix?.map?.spawnGates ?? []).filter((gate) => gate.edge === edge && typeof gate.x === "number" && typeof gate.z === "number");
  if (!gates.length) return null;
  const center = gates.reduce((acc, gate) => ({ x: acc.x + gate.x / gates.length, z: acc.z + gate.z / gates.length }), { x: 0, z: 0 });
  const dx = center.x - (from.x ?? 0);
  const dz = center.z - (from.z ?? 0);
  const d = Math.hypot(dx, dz);
  if (!d) return null;
  return { x: dx / d, z: dz / d };
}

File: artifacts/gauntlet-heat2-20260824/hermes/rider_driver.py
#!/usr/bin/env python3
import argparse
import json
import subprocess
import sys
import time
from pathlib import Path


def pos_ring(claim):
    x, z = claim["x"], claim["z"]
    return [
        (x, z - 5), (x - 5, z), (x + 5, z), (x, z + 5),
        (x - 4, z - 4), (x + 4, z - 4), (x - 4, z + 4), (x + 4, z + 4),
        (x, z - 8), (x - 8, z), (x + 8, z), (x, z + 8),
    ]


def choose_build(contract, mechanics, now, claim):
    available = {b["id"]: b for b in mechanics.get("buildables", [])}
    counts = now.get("works", {}).get("byKind", {})
    standing = [e for e in now.get("works", {}).get("entries", []) if not e.get("wrecked")]
    occupied = [(e.get("position", {}).get("x"), e.get("position", {}).get("z")) for e in now.get("works", {}).get("entries", [])]

    if contract == "the-claim":
        plan = ["palisade", "palisade", "sentry_beacon", "palisade", "palisade", "turret", "sentry_beacon", "palisade", "turret", "palisade"]
    elif contract == "e1-night-shift":
        palisades = int(counts.get("palisade", 0))
        beacons = int(counts.get("sentry_beacon", 0))
        turrets = int(counts.get("turret", 0))
        if now.get("gold", 0) >= 50 and turrets < 2:
            wanted = "turret"
        elif palisades < 4:
            wanted = "palisade"
        elif beacons < 1:
            wanted = "sentry_beacon"
        else:
            wanted = "palisade"
        plan = [wanted]
    else:
        palisades = int(counts.get("palisade", 0))
        beacons = int(counts.get("sentry_beacon", 0))
        turrets = int(counts.get("turret", 0))
        if palisades < 4:
            wanted = "palisade"
        elif now.get("gold", 0) >= 50 and turrets < 3:
            wanted = "turret"
        elif beacons < 1:
            wanted = "sentry_beacon"
        else:
            wanted = "palisade"
        plan = [wanted]

    built_total = len(standing)
    wanted = plan[min(built_total, len(plan) - 1)]
    if wanted not in available:
        for fallback in ("turret", "sentry_beacon", "palisade", "boiler_house", "lantern_post"):
            if fallback in available:
                wanted = fallback
                break
        else:
            return None

    b = available[wanted]
    count = int(counts.get(wanted, 0)) if isinstance(counts, dict) else 0
    costs = b.get("costs", [b.get("cost", 0)])
    cost = costs[min(count, len(costs) - 1)]

    active_seams = [s for s in now.get("seams", []) if s.get("active")]
    projected = now.get("gold", 0) + 5 * len(active_seams)
    if projected < cost:
        return None

    if contract == "e2-hill-mine":
        x, z = claim["x"], claim["z"]
        ring = [(x, z + 4), (x - 4, z + 4), (x + 4, z + 4), (x, z + 8), (x - 6, z + 7), (x + 6, z + 7)]
    else:
        ring = pos_ring(claim)
    for px, pz in ring:
        if all(ox is None or (px - ox) ** 2 + (pz - oz) ** 2 >= 9 for ox, oz in occupied):
            return wanted, cost, {"x": px, "z": pz}
    return None


def orders_for(view):
    stable = view["stablePrefix"]
    now = view["now"]
    contract = stable["contract"]["id"]
    mechanics = stable["mechanics"]
    claim = stable["map"]["claim"]

    if now.get("pendingSecure"):
        return [{"verb": "SECURE_CHOICE", "choice": "bank"}]
    if now.get("pendingOffer"):
        offer = now["pendingOffer"]
        preferred = offer[0]
        for choice in offer:
            text = (choice.get("name", "") + " " + choice.get("effectText", "")).lower()
            if any(k in text for k in ("max hp", "heals", "health", "repair")):
                preferred = choice
                break
        else:
            for choice in offer:
                text = (choice.get("name", "") + " " + choice.get("effectText", "")).lower()
                if any(k in text for k in ("damage", "fire rate", "spark", "blast", "range")):
                    preferred = choice
                    break
        return [{"verb": "PICK_UPGRADE", "id": preferred["id"]}]

    orders = [{"verb": "SET_WEAPON", "weapon": "blast" if contract == "e2-hill-mine" else "rig"}]
    if now.get("threats", {}).get("alive", 0) and now.get("blastReadyInMs") == 0:
        hero = now.get("hero", {})
        tx, tz = hero.get("x", claim["x"]), hero.get("z", claim["z"])
        edge = now.get("threats", {}).get("edge")
        if edge == "north":
            tz += 8
        elif edge == "south":
            tz -= 8
        elif edge == "east":
            tx += 8
        elif edge == "west":
            tx -= 8
        orders.append({"verb": "BLAST_AT", "pos": {"x": tx, "z": tz}})
    if now.get("needsRider") and now.get("threats", {}).get("alive", 0):
        orders.append({"verb": "HOLD", "pos": claim})
        return orders
    active_seams = [s for s in now.get("seams", []) if s.get("active")]
    active_seams.sort(key=lambda s: (s.get("x", claim["x"]) - now.get("prospector", {}).get("x", claim["x"])) ** 2 + (s.get("z", claim["z"]) - now.get("prospector", {}).get("z", claim["z"])) ** 2)
    for seam in active_seams:
        orders.append({"verb": "HARVEST", "seam": seam["id"]})

    build = choose_build(contract, mechanics, now, claim)
    orders.append({"verb": "MOVE_TO", "pos": claim})
    if build:
        what, cost, where = build
        orders.append({"verb": "BUILD", "what": what, "where": where, "when": {"goldGte": cost}})
    orders.append({"verb": "REPAIR_UNDER", "pct": 75})
    orders.append({"verb": "HOLD", "pos": claim})
    return orders[:32]


def summary(view):
    now = view["now"]
    return {
        "wave": now.get("wave"),
        "gold": now.get("gold"),
        "heroHp": now.get("hero", {}).get("hp"),
        "prospector": now.get("prospector"),
        "works": now.get("works"),
        "threats": now.get("threats"),
        "seams": now.get("seams"),
        "offer": now.get("pendingOffer"),
        "secure": now.get("pendingSecure"),
        "needsRider": now.get("needsRider"),
        "lastLog": view.get("appendLog", [])[-1:] if view.get("appendLog") else [],
    }


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--contract", required=True)
    ap.add_argument("--seed", required=True)
    ap.add_argument("--tape", required=True)
    ap.add_argument("--transcript", required=True)
    args = ap.parse_args()
    cmd = ["node", "scripts/gr-sim.mjs", "--contract", args.contract, "--seed", args.seed, "--tape", args.tape]
    start = time.monotonic()
    with Path(args.transcript).open("w") as log:
        proc = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, bufsize=1)
        outcome = None
        assert proc.stdout and proc.stdin
        for line in proc.stdout:
            log.write("VIEW " + line)
            log.flush()
            obj = json.loads(line)
            if obj.get("schema") != "goldrush.view.v1":
                outcome = obj
                print(json.dumps({"outcome": obj, "wallSeconds": round(time.monotonic() - start, 3)}), flush=True)
                break
            print(json.dumps(summary(obj), separators=(",", ":")), flush=True)
            orders = orders_for(obj)
            log.write("ORDERS " + json.dumps(orders, separators=(",", ":")) + "\n")
            log.flush()
            proc.stdin.write(json.dumps(orders, separators=(",", ":")) + "\n")
            proc.stdin.flush()
        stderr = proc.stderr.read() if proc.stderr else ""
        code = proc.wait()
        log.write("STDERR\n" + stderr)
        log.write(f"\nEXIT {code}\nWALL {time.monotonic() - start:.3f}\n")
        if code != 0 or outcome is None:
            print(json.dumps({"exit": code, "stderr": stderr[-2000:], "wallSeconds": round(time.monotonic() - start, 3)}), flush=True)
            sys.exit(code or 2)


if __name__ == "__main__":
    main()
